distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso

```diff
@@ -1083,6 +1089,158 @@ type Table
 
+    ## ALIAS append, concat
+       Appends records from other table(s) to this table.
+
+       Arguments:
+       - tables: A single table or a vector of tables to append to this one. The
+         tables are concatenated in the order they are specified, with `self`
+         being the first one.
+       - match_columns: Specifies how to match the columns.
+         - If `Match_Columns.By_Name` - the columns are matched by name across
+           all provided tables.
+           If unmatched columns are to be dropped, the resulting table will keep
+           only the set of columns that appear in all provided tables, in the
+           relative order that they appeared in the `self` table.
+           If unmatched columns are kept, they are added in the order of
+           appearance - i.e. first all columns from `self` will be added in the
+           original order, then any columns from the second table that were not
+           matched will be added at the end (preserving their relative order),
+           and so on for all the remaining tables.
+         - If `Match_Columns.By_Position` - the columns are mapped by position.
+           If unmatched columns are to be dropped, the resulting table will have
+           as many columns as the table that had the least columns and the
+           column names of the first table (self) will be used.
+           If unmatched columns are kept, the resulting table will have as many
+           columns as the table with the most columns. Since the first table may
+           not have all the necessary columns to provide column names for the
+           result, the result will have column names taken from the first table
+           that has the biggest number of columns.
+       - keep_unmatched_columns: If set to `True`, unmatched columns are kept
+         and are padded with `Nothing` for tables that did not have them.
+         If set to `False`, only the common subset of columns is kept - any
+         column that is not present in all tables is dropped. Defaults to
+         `Report_Unmatched`, which behaves like `True` - unmatched columns are
+         kept and padded with `Nothing`, but a problem is reported.
+       - allow_type_widening: Specifies if the resulting column type should be
+         adapted to fit columns from all arguments; defaults to `True`. If set
+         to `True` a common type will be chosen for each column, as described in
+         the section "Unifying Column Types" below. If set to `False` the
+         resulting column type will be the same as in the first table the
+         corresponding column appeared in (usually `self`, except for unmatched
+         columns). If type widening is not allowed, all columns that are merged
+         with each other must have the same type as the first one. The only
+         exception is if the first column had `Mixed` type which will accept any
+         other type.
```

Comment:
I've updated the behaviour so that if `allow_type_widening=False`, then _all_ corresponding columns must have the same type.

The only exception is if the first column has type `Mixed`, then any type is allowed for others.

Does that sound good @jdunkerley ? It is a slight detour from the original design that we agreed on, but the idea of having the special handling for int->float widening was IMO quite 'random' - so to make stuff consistent I decided to handle all the same. The only exception is `Mixed` - I'm still on the fence there - `int -> mixed` is _still_ _some_ kind of widening so I'm not sure.

Maybe if I disable all widening, I should not even allow widening a specific type to the more generic one `Mixed`? Then this would be a strict mode that will just check that all column types match. I use `allow_type_widening=False` for a reason, after all, since the default one is allowing widening.

Still, allowing this exception when the first column is Mixed sounds rather practical - I usually want `allow_type_widening` to keep the same column types as in the first table and this is still achieved - the type is still `Mixed`.

I'm slowly leaning towards disallowing even this exception, but I think I'd like to hear a second opinion first.

I'm open to discussion here - whichever approach is easy to switch as once we have the type logic changing this detail is rather simple. I'd like this to be as intuitive and useful for users as possible.

Reply:
I feel like this uncertainty is actually stemming from `allow_type_widening=False` serving **two/three** kind of separate purposes.

1. It ensures that all columns in the output will have the type as the first table that contained that column had - so the type of the first table's column will not change.
2. It ensures that types of contained data do not change, especially when they could lose precision (like a long->double conversion may lose precision for large numbers IIRC).
3. It acts like a check that all merged columns actually had the same type.

The (1) and (2) are in favour of allowing the Mixed coercion, (3) is not.
If we want just (1), we could also allow the bool->int and int->float coercions if the first column is int and float respectively, as they will still make (1) hold. But these are breaking (2) and (3) so I think they should be removed, as I did (especially as we had just int->float, but not bool->int which was inconsistent).

Still, I'm thinking if the (3) argument is not sensible and we should remove the Mixed exception too. Any thoughts?

Reply:
Honestly, I think both are viable usecases for users and both can be checked in other ways, so whichever we choose should be ok. I think we may want to think what is the more common usecase. I like (3) for being more correctness-oriented, but in data-science usecases (1) seems like the most important one and (2) may be useful.

So this argumentation leads me to believe that the current solution is just right. Do you agree?

Reply:
I think the current implementation feels about correct.

I expect we will revisit this as we do the Value Types, but this seems sensible to me, especially if we error clearly and allow users to easily re-cast the columns as needed.